From daf0b5579965bfca231b57e28b9c6f2d48a320e1 Mon Sep 17 00:00:00 2001
From: Justin Zandbergen <justin.zandbergen@topicus.nl>
Date: Fri, 23 Feb 2024 10:50:22 +0100
Subject: [PATCH 1/4] fix: Older keys end with a newline, this breaks minion
 auth.

---
 salt/channel/server.py | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/salt/channel/server.py b/salt/channel/server.py
index 4d7cd23fc29..ba577c3dadb 100644
--- a/salt/channel/server.py
+++ b/salt/channel/server.py
@@ -371,7 +371,15 @@ def _auth(self, load, sign_messages=False):
         elif os.path.isfile(pubfn):
             # The key has been accepted, check it
             with salt.utils.files.fopen(pubfn, "r") as pubfn_handle:
-                if salt.crypt.clean_key(pubfn_handle.read()) != load["pub"]:
+                keyFromDisk = pubfn_handle.read()
+
+                # if the keyFromDisk has a final newline it is a oldstyle key
+                # if we clean it, it will not match. Only clean the key if it
+                # is a new style key.
+                if keyFromDisk[-1:] != "\n":
+                    keyFromDisk = salt.crypt.clean(orgkey)
+
+                if  keyFromDisk != load["pub"]:
                     log.error(
                         "Authentication attempt from %s failed, the public "
                         "keys did not match. This may be an attempt to compromise "

From f69f4cd1ac75c2f13ebf2ad671cfc95966af3e59 Mon Sep 17 00:00:00 2001
From: Justin Zandbergen <justin.zandbergen@topicus.nl>
Date: Fri, 23 Feb 2024 11:04:31 +0100
Subject: [PATCH 2/4] fix: typo's

---
 salt/channel/server.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/salt/channel/server.py b/salt/channel/server.py
index ba577c3dadb..68c3005d685 100644
--- a/salt/channel/server.py
+++ b/salt/channel/server.py
@@ -377,7 +377,7 @@ def _auth(self, load, sign_messages=False):
                 # if we clean it, it will not match. Only clean the key if it
                 # is a new style key.
                 if keyFromDisk[-1:] != "\n":
-                    keyFromDisk = salt.crypt.clean(orgkey)
+                    keyFromDisk = salt.crypt.clean_key(keyFromDisk)

                 if  keyFromDisk != load["pub"]:
                     log.error(

From 7469a705a63e34dcec816da8a3937f92d2054428 Mon Sep 17 00:00:00 2001
From: Justin Zandbergen <justin.zandbergen@topicus.nl>
Date: Mon, 26 Feb 2024 10:09:46 +0100
Subject: [PATCH 3/4] Simply check against cleaned key from disk.

---
 salt/channel/server.py | 10 +---------
 1 file changed, 1 insertion(+), 9 deletions(-)

diff --git a/salt/channel/server.py b/salt/channel/server.py
index 68c3005d685..79660720485 100644
--- a/salt/channel/server.py
+++ b/salt/channel/server.py
@@ -371,15 +371,7 @@ def _auth(self, load, sign_messages=False):
         elif os.path.isfile(pubfn):
             # The key has been accepted, check it
             with salt.utils.files.fopen(pubfn, "r") as pubfn_handle:
-                keyFromDisk = pubfn_handle.read()
-
-                # if the keyFromDisk has a final newline it is a oldstyle key
-                # if we clean it, it will not match. Only clean the key if it
-                # is a new style key.
-                if keyFromDisk[-1:] != "\n":
-                    keyFromDisk = salt.crypt.clean_key(keyFromDisk)
-
-                if  keyFromDisk != load["pub"]:
+                if salt.crypt.clean_key(pubfn_handle.read()) != salt.crypt.clean_key(load["pub"]):
                     log.error(
                         "Authentication attempt from %s failed, the public "
                         "keys did not match. This may be an attempt to compromise "

From 90a5ce05fda6379db088e2ac8b19fef25b49caef Mon Sep 17 00:00:00 2001
From: Justin Zandbergen <justin.zandbergen@topicus.nl>
Date: Mon, 26 Feb 2024 10:40:10 +0100
Subject: [PATCH 4/4] Add changelog entry

---
 changelog/66126.fixed.md | 2 ++
 1 file changed, 2 insertions(+)
 create mode 100644 changelog/66126.fixed.md

diff --git a/changelog/66126.fixed.md b/changelog/66126.fixed.md
new file mode 100644
index 00000000000..9879189e644
--- /dev/null
+++ b/changelog/66126.fixed.md
@@ -0,0 +1,2 @@
+Fixed a issue with server channel where a minion's public key
+would be rejected if it contained a final newline character.
